File: scripts/webos/gen_i18n.py
#!/usr/bin/env python3
import json
import re
from argparse import ArgumentParser
from os import path
from pathlib import Path
from typing import Dict, List

import polib


class Args:
    output: str
    locales: List[str]


parser = ArgumentParser()
parser.add_argument('-o', '--output', required=True)
parser.add_argument('locales', metavar='LOCALE', type=str, nargs='+')

args: Args = parser.parse_args(namespace=Args())

outdir = args.output
locales = args.locales

for locale in locales:
    popath = Path('i18n', locale, 'messages.po')
    pofile: polib.POFile = polib.pofile(str(popath))
    cstrings: Dict[str, str] = {}
    for entry in pofile.translated_entries():
        cstrings[entry.msgid] = entry.msgstr
    lang_segs = re.compile(r'[-_]').split(locale)
    lang_dir = Path(args.output, *lang_segs)
    if not lang_dir.exists():
        lang_dir.mkdir(parents=True, exist_ok=True)
    print(f'{popath} => {path.join(*lang_segs, "cstrings.json")}')
    with lang_dir.joinpath('cstrings.json').open('w') as f:
        json.dump(cstrings, f, ensure_ascii=True)

File: scripts/webos/gen_i18n.js
#!/usr/bin/env node
'use strict';
const {po} = require('gettext-parser');
const {ArgumentParser} = require('argparse');
const path = require('path');
const fs = require('fs');

const parser = new ArgumentParser()
parser.add_argument('-o', '--output', {required: true});
parser.add_argument('locales', {metavar: 'LOCALE', type: String, nargs: '+'})

const args = parser.parse_args();

const outdir = args['output'];
const locales = args['locales'];

for (const locale of locales) {
    let cstrings = {}
    const popath = path.join('i18n', locale, 'messages.po');
    const pofile = po.parse(fs.readFileSync(popath));
    const translations = pofile.translations[''];
    for (const msgid in translations) {
        if (!msgid) continue;
        const translation = translations[msgid];
        const msgstr = translation.msgstr && translation.msgstr[0];
        if (!msgstr) continue;
        cstrings[translation.msgid] = msgstr;
    }

    const language = locale.split(/[-_]/);
    const lang_path = path.join(outdir, ...language);
    if (!fs.existsSync(lang_path)) {
        fs.mkdirSync(lang_path, {recursive: true});
    }
    console.log(`${popath} => ${path.join(...language, 'cstrings.json')}`);
    fs.writeFileSync(path.join(lang_path, 'cstrings.json'), JSON.stringify(cstrings));
}